### DISCUSSION TITLE
automation with pipeline/terraform

### DISCUSSION THREAD

**Discussion by gksetu23** [gravitational/teleport#30942] [Q&A]
Context: Need to automate management of users/roles to give access to people on databases.

Hi Team, need some help on how I can automate management of resources like (user, roles) via terraform that to be integrated with pipeline. 

To have it run with terraform, is there any way to provide direct authentication/token instead of identity file for which it needs to have access for the teleport auth server of a user and then it needs 2F auth which is not very convenient to use with automation setup.

In ref to this doc for impersonate a user - https://goteleport.com/docs/access-controls/guides/impersonation/

**Comment by webvictim** [comment]
Take a look at this guide for using the [Teleport Terraform provider](https://goteleport.com/docs/management/dynamic-resources/terraform-provider) with [Machine ID](https://goteleport.com/docs/machine-id/introduction/): https://github.com/gravitational/teleport/discussions/28491

**Reply by gksetu23** [reply]
@webvictim thanks for sharing that, but the main issue is automation where it'll be running (some pipeline container) - how does it get access to `tsh/tctl` (teleport cluster) in the first place? As login requires user-password method and it requires OTP too. Is there a better way to get access direct to the auth cluster so that it'll be straight forward to use with terraform

**Reply by webvictim** [reply]
If you take a look at the [Machine ID guides](https://goteleport.com/docs/machine-id/guides/) page, there's a guide for Gitlab (https://goteleport.com/docs/machine-id/guides/gitlab/) that describes quite a similar process.  Essentially if you use a supported platform you can grant bots access to join the cluster from a trusted pipeline or provider.

This is a little more difficult when using standard join tokens, but the general idea behind Machine ID is that you can delegate a role or identity to a bot or machine service so that `tctl` access isn't needed, and nor is a password or MFA like with a regular user.

**Reply by gksetu23** [reply]
Got it @webvictim, I did give it a try to test it out but ran into some issues when testing out with the workflow. Can you help with the value for this  var `"TBOT_GITLAB_JWT"` as mentioned in the below document. (I kept is as `"gitlab-demo"` as thats what the `tctl bots add ..` gave.)

ref: https://goteleport.com/docs/machine-id/guides/gitlab/

error logs for workflow:

<img width="1184" alt="image" src="https://github.com/gravitational/teleport/assets/143083224/f216f411-2812-4b36-aa55-b9cbd8e1bacb">

**Reply by irbis1q38** [reply]
@gksetu23 
Have you managed to beat this error? Have the same issue...
